File: ABC097/D_Equals/submission2.java
// AC

import java.util.Scanner;

class Main {
  public static int root(int[] root, int i) {
    int r = root[i];
    if (r == i) {
      return r;
    } else {
      int r2 = root(root, r);
      root[i] = r2;
      return r2;
    }
  }

  public static void main(String[] args) {

    var sc = new Scanner(System.in);

    int n = sc.nextInt();
    int m = sc.nextInt();

    var ps = new int[n];
    for (int i = 0; i < n; i++) {
      ps[i] = sc.nextInt() - 1;
    }

    var ab = new int[2 * m];
    for (int i = 0; i < m; i++) {
      int a = sc.nextInt() - 1;
      int b = sc.nextInt() - 1;
      ab[2 * i    ] = a;
      ab[2 * i + 1] = b;
    }

    var root = new int[n];
    for (int i = 0; i < n; i++) {
      root[i] = i;
    }
    for (int i = 0; i < m; i++) {
      int a = ab[2 * i    ];
      int b = ab[2 * i + 1];
      int ra = root(root, a);
      int rb = root(root, b);
      if (ra != rb) {
        root[rb] = ra;
      }
    }

    int answer = 0;
    for (int i = 0; i < n; i++) {
      int p = ps[i];
      int rp = root(root, p);
      int ri = root(root, i);
      if (ri == rp) {
        answer++;
      }
    }

    System.out.println(answer);

  }
}

File: ABC138/D_Ki/submission2.scala
// AC

object Main extends App {
  val sc = new java.util.Scanner(System.in);
  val n, q = sc.nextInt();
  val abs = Array.fill(n - 1)((sc.nextInt() - 1, sc.nextInt() - 1));
  val pxs = Array.fill(q)((sc.nextInt() - 1, sc.nextInt()));

  val abs2 = Array.fill[List[Int]](n)(Nil);
  abs.foreach { case (a, b) =>
    abs2(a) = b :: abs2(a);
    abs2(b) = a :: abs2(b);
  }

  val answer = Array.fill(n)(0);
  pxs.foreach { case (p, x) =>
    answer(p) += x;
  }
  @scala.annotation.tailrec
  def sub(stack: List[(Int, List[Int])]): Unit = {
    stack match {
      case Nil => ;
      case (_, Nil) :: stack2 => sub(stack2);
      case (parent, h :: tail) :: stack2 =>
        val cs = abs2(h).filter(_ != parent);
        if (!cs.isEmpty) {
          cs.foreach { c =>
            answer(c) += answer(h);
          }
          sub((h, cs) :: (parent, tail) :: stack2);
        } else if (!tail.isEmpty) {
          sub((parent, tail) :: stack2);
        } else {
          sub(stack2);
        }
    }
  }
  sub((-1, List(0)) :: Nil);

  println(answer.mkString(" "));
}
